File: sdk/java/src/main/java/com/xxxxteam/tempmail/providers/TempFastMail.java
package com.xxxxteam.tempmail.providers;

import com.google.gson.JsonElement;
import com.xxxxteam.tempmail.Email;
import com.xxxxteam.tempmail.EmailInfo;
import com.xxxxteam.tempmail.HttpClient;
import com.xxxxteam.tempmail.HttpResult;
import com.xxxxteam.tempmail.Json;
import com.xxxxteam.tempmail.Normalizer;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * TempFastMail 渠道 — https://tempfastmail.com
 * POST /api/email-box 创建，GET /api/email-box/{uuid}/emails 收信。
 */
public final class TempFastMail {

    private static final String BASE_URL = "https://tempfastmail.com";
    private static final Map<String, String> HEADERS = new LinkedHashMap<>();

    static {
        HEADERS.put("Accept", "application/json");
        HEADERS.put("User-Agent", "Mozilla/5.0");
    }

    private TempFastMail() {
    }

    /**
     * 创建临时邮箱。
     *
     * @return 邮箱信息，token 为 uuid
     */
    public static EmailInfo generate() {
        HttpResult resp = HttpClient.post(BASE_URL + "/api/email-box", null, null, HEADERS);
        resp.ensureSuccess();
        JsonElement data = Json.parse(resp.getBody());
        String email = Json.str(data, "email").trim();
        String uuid = Json.str(data, "uuid").trim();
        if (email.isEmpty() || !email.contains("@") || uuid.isEmpty()) {
            throw new RuntimeException("tempfastmail: 创建邮箱响应无效");
        }
        return new EmailInfo("tempfastmail", email, uuid, null, null);
    }

    /**
     * 获取邮件列表。
     *
     * @param token uuid
     * @param email 邮箱地址
     * @return 邮件列表
     */
    public static List<Email> getEmails(String token, String email) {
        String uuid = (token != null ? token : "").trim();
        String address = (email != null ? email : "").trim();
        if (uuid.isEmpty()) {
            throw new RuntimeException("tempfastmail: token 为空");
        }
        if (address.isEmpty()) {
            throw new RuntimeException("tempfastmail: 邮箱地址为空");
        }
        HttpResult resp = HttpClient.get(
                BASE_URL + "/api/email-box/" + ProviderUtil.urlEncode(uuid) + "/emails", HEADERS);
        resp.ensureSuccess();
        JsonElement parsed = Json.parse(resp.getBody());
        List<Email> result = new ArrayList<>();
        if (parsed == null || !parsed.isJsonArray()) {
            return result;
        }
        for (JsonElement item : parsed.getAsJsonArray()) {
            if (item == null || !item.isJsonObject()) {
                continue;
            }
            String msgUuid = Json.str(item, "uuid").trim();
            JsonElement detail = null;
            if (!msgUuid.isEmpty()) {
                detail = fetchDetail(uuid, msgUuid);
            }
            Map<String, Object> raw = buildRaw(detail != null ? detail : item, address);
            result.add(Normalizer.normalizeEmail(raw, address));
        }
        return result;
    }

    private static JsonElement fetchDetail(String boxUuid, String msgUuid) {
        try {
            HttpResult resp = HttpClient.get(
                    BASE_URL + "/api/email-box/" + ProviderUtil.urlEncode(boxUuid)
                            + "/email/" + ProviderUtil.urlEncode(msgUuid), HEADERS);
            if (!resp.isOk()) {
                return null;
            }
            return Json.parse(resp.getBody());
        } catch (Exception e) {
            return null;
        }
    }

    private static Map<String, Object> buildRaw(JsonElement el, String recipient) {
        Map<String, Object> raw = Json.toDict(el);
        raw.put("id", raw.getOrDefault("uuid", ""));
        raw.putIfAbsent("to", raw.getOrDefault("real_to", recipient));
        raw.putIfAbsent("date", raw.getOrDefault("received_at", ""));
        return raw;
    }
}

File: sdk/kotlin/src/main/kotlin/com/xxxxteam/tempmail/providers/TempFastMail.kt
package com.xxxxteam.tempmail.providers

import com.xxxxteam.tempmail.Email
import com.xxxxteam.tempmail.EmailInfo
import com.xxxxteam.tempmail.Http
import com.xxxxteam.tempmail.Normalize
import com.xxxxteam.tempmail.Provider
import kotlinx.serialization.json.*

/**
 * TempFastMail 渠道（tempfastmail.com）。
 *
 * POST /api/email-box 创建，token 为 uuid；GET /api/email-box/{uuid}/emails 收信。
 */
object TempFastMail : Provider {

    private const val BASE_URL = "https://tempfastmail.com"
    private val HEADERS = mapOf(
        "Accept" to "application/json",
        "User-Agent" to "Mozilla/5.0",
    )

    /** 创建临时邮箱，token 为 uuid。 */
    override suspend fun generate(): EmailInfo {
        val resp = ProviderUtil.httpPost("$BASE_URL/api/email-box", null, null, HEADERS)
        resp.ensureSuccess()
        val data = Http.json.parseToJsonElement(resp.body) as? JsonObject
            ?: throw RuntimeException("tempfastmail: 创建邮箱响应无效")
        val email = jsonStr(data, "email").trim()
        val uuid = jsonStr(data, "uuid").trim()
        if (email.isEmpty() || !email.contains("@") || uuid.isEmpty()) {
            throw RuntimeException("tempfastmail: 创建邮箱响应无效")
        }
        return EmailInfo(email = email, channel = "tempfastmail", token = uuid)
    }

    /** 获取邮件列表。 */
    override suspend fun getEmails(info: EmailInfo): List<Email> {
        val uuid = info.token.trim()
        val address = info.email.trim()
        if (uuid.isEmpty()) throw RuntimeException("tempfastmail: token 为空")
        if (address.isEmpty()) throw RuntimeException("tempfastmail: 邮箱地址为空")
        val resp = ProviderUtil.httpGet("$BASE_URL/api/email-box/${ProviderUtil.urlEncode(uuid)}/emails", HEADERS)
        resp.ensureSuccess()
        val arr = Http.json.parseToJsonElement(resp.body) as? JsonArray ?: return emptyList()
        return arr.mapNotNull { it as? JsonObject }.map { item ->
            val msgUuid = jsonStr(item, "uuid").trim()
            val src = if (msgUuid.isNotEmpty()) fetchDetail(uuid, msgUuid) ?: item else item
            Normalize.fromJson(buildRaw(src, address), address)
        }
    }

    /** 拉取单封邮件详情。 */
    private suspend fun fetchDetail(boxUuid: String, msgUuid: String): JsonObject? {
        return try {
            val resp = ProviderUtil.httpGet(
                "$BASE_URL/api/email-box/${ProviderUtil.urlEncode(boxUuid)}/email/${ProviderUtil.urlEncode(msgUuid)}",
                HEADERS,
            )
            if (!resp.isOk) return null
            Http.json.parseToJsonElement(resp.body) as? JsonObject
        } catch (_: Exception) {
            null
        }
    }

    /** 覆盖 id/to/date 后交由 Normalize 做候选字段提取。 */
    private fun buildRaw(obj: JsonObject, recipient: String): JsonObject = buildJsonObject {
        obj.forEach { (k, v) -> put(k, v) }
        put("id", jsonStr(obj, "uuid"))
        if (obj["to"] == null) put("to", ProviderUtil.firstNonEmpty(jsonStr(obj, "real_to"), recipient))
        if (obj["date"] == null) put("date", jsonStr(obj, "received_at"))
    }
}

/** 从 JSON 对象读取字符串字段，缺失返回空串（文件私有）。 */
private fun jsonStr(obj: JsonObject, key: String): String =
    (obj[key] as? JsonPrimitive)?.contentOrNull ?: ""
